spack logs: added new command

alalazo opened the pull request:
Added a new command `spack log show`, that mimics `docker logs`:
```console
$ spack log show libszip
#
# ... dumps the file ...
#
```
It takes a constraint that identifies a unique spec, and dumps its `build.out`. The log can be piped to other linux commands (such as `grep` or less). Errors are handled gracefully:
```console
$ spack log show foobar
==> Error: Spec 'foobar' matches no installed packages.

$ spack log show foobar foo
==> Error: only one spec is allowed in the query [2 given]

$ spack log show %gcc
==> Error: %gcc matches multiple packages.
  Matching packages:
    tfkam54 bison@3.0.4%gcc@4.8 arch=linux-ubuntu14.04-x86_64 
    ...
    eksallf zlib@1.2.11%gcc@4.8 arch=linux-ubuntu14.04-x86_64 
  Use a more specific spec.
```

The command `spack log-parse` has been moved to `spack log parse` and now accepts either a file or a spec:
```console
$ spack log parse -s zlib
0 errors

$ spack log parse -s tcl
1 errors
     655    checking for 64-bit integer type... using long
     656    checking for build with symbols... no
     657    checking for tclsh... /tmp/pytest-of-mculpo/pytest-0/test_keep_exceptions0/tmp/spack-stage/spack-stage-aUKyP3/tcl8.6.6/unix/tclsh
  >> 658    /tmp/pytest-of-mculpo/pytest-0/test_keep_exceptions0/tmp/spack-stage/spack-stage-aUKyP3/tcl8.6.6/pkgs/tdbc1.0.4/configure: line 8999: cd: /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.
            04-x86_64/gcc-4.8/tcl-8.6.6-xrem43kcgm3ij2drw3vqhnxvhzi3yonf/lib/tdbc1.0.4: No such file or directory
     659    configure: creating ./config.status
     660    config.status: creating Makefile
     661    config.status: creating pkgIndex.tcl
```

tgamblin commented on lib/spack/spack/cmd/logs.py:
```diff
@@ -0,0 +1,85 @@
+        tty.die(msg.format_map(spec.short_spec))
+
+    # Construct the filename to be opened
+    filename = 'build.env' if args.environment else 'build.out'
```
why not make several arguments that use `store_const` with these string values and a single `filename` as `dest`?  Then the parser will build the name for you.

tgamblin commented on lib/spack/spack/cmd/logs.py:
```diff
@@ -0,0 +1,85 @@
+
+def setup_parser(subparser):
+    subparser.add_argument(
+        '--environment',
```
I would call this `--build-env` or `--build-environment`.  I'm somewhat conflicted about whether this should be an argument to this command or to `spack env`, e.g. `spack env --installed <spec>`.  The latter would also let you actually run a command in the old environment, which could be very useful.

alalazo replied:
Let me know if you want this command at all, or if you want just want 
```
spack env --install <spec>
```

For the what is worth, the motivation that lead me putting this PR together was just being tired to repeat:
```
$ spack find -p <spec>
==> 1 installed packages.
-- linux-ubuntu14.04-x86_64 / gcc@4.8 ---------------------------
    zlib@1.2.11  /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/zlib-1.2.11-eksallf6cymqkp6pkz6ymzjakqt6bqkx

$ less <cut'n paste of the path above>/.spack/build.out
```
over and over when looking at previous installations logs. Having a dump of the environment was just a low hanging fruit once opening the logs had been implemented.

tgamblin replied:
How about just these:

* `spack env --install <spec>`
* `spack edit --log <spec>`

or just the latter if you're short on time?  I think the use case you describe is a good one, we just already have `spack edit`

tgamblin replied:
PS -- `spack edit` probably only need another arg and a call to `disambiguate spec` in `edit_package` to make this work

tgamblin commented on lib/spack/spack/cmd/logs.py:
```diff
@@ -0,0 +1,85 @@
+        msg = 'log file does not exist! [{0}]'
+        tty.die(msg.format(abs_filename))
+
+    spack.editor(abs_filename)
```
I think if this is just going to edit the file, it should be implemented as an argument to `spack edit`.  e.g., `spack edit --build-log <spec>`.

Similarly, if it is going to print the filename, there should probably be an argument to `spack location`.

What we don't have currently is something to just dump the log, so maybe this command should just do that?  Seems like it should also be called `spack show-log` if it does that.  But see my comments about about overlap with `spack env` for the environment log.

alalazo replied:
`spack.editor` was the fastest way I could think of to open the logs into something paged. 

I would really like to be able to somehow open the file read only, but couldn't find a decent way to do it that works for every editor I am aware of. An alternative, to avoid accidental corruptions of the logs, is to copy them in some `tmp` space and open the copy + delete it on editor exit.

Note that if you want just a dump, for instance within a script, you can:
```
export EDITOR=cat 
spack logs <spec>
```
Does this seem reasonable to you?

tgamblin replied:
I guess it seems kind of awkward?  What about just:

```
spack show-log <spec>
```

alalazo commented on lib/spack/spack/cmd/logs.py:
```diff
@@ -0,0 +1,85 @@
+        spack.cmd.display_specs(specs, args)
+        raise SystemExit(1)
+
+    spec = specs[0]
```
Thanks for the tip.

alalazo commented:
@tgamblin Just dumping the log right now. Will do a PR for:
```
spack env --install <spec>
```
later on. Also, not sure if in the end you wanted:
```
spack show-log <spec>
```
or directly a sub-command:
```
spack log show <spec>
```
Just post your preferred, I'll change the naming here.

alalazo commented:
ping

alalazo commented:
> I think it would be useful if it dumped the log, as that would allow someone to grep for errors.

I think I addressed the comment. Let me know if there's something else to do here.

tgamblin commented:
@alalazo I kind of like `spack log show` here.  I think that gives us some flexibility for adding new stuff.

Speaking of which, see #7093.  Should that be added here?  The only thing is that `spack log-parse` there takes a file, not a spec.  Maybe it should be merged with this as

```
spack log parse [-f file | spec]
```

alalazo commented:
@tgamblin I agree with merging these two as:
```console
spack log parse [-f file | spec]
spack log show [-f file | spec]
```
Ok if I proceed with that in this PR?

tgamblin commented:
sure!  go ahead

alalazo commented:
@tgamblin Ready for another review!

alalazo commented:
ping
